compiler/src/codegen/compcore.re

```diff
@@ -2935,9 +2935,151 @@ and compile_instr = (wasm_mod, env, instr) =>
   | MTagOp(_) => failwith("NYI: (compile_instr): MTagOp")
   };
 
+let runtime_type_metadata_ptr = 256;
+
+let compile_type_metadata = (wasm_mod, env, type_metadata) => {
+  open Types;
+
+  // More information about this function can be found in the printing.md
+  // contributor document.
+
+  // Extension constructors defined by the module must be grouped together.
+  // For now, this only includes exceptions.
+  let (exception_meta, non_exception_meta) =
+    List.partition(
+      fun
+      | ExceptionMetadata(_) => true
+      | _ => false,
+      type_metadata,
+    );
+  let type_metadata =
+    switch (exception_meta) {
+    | [ExceptionMetadata(id, _, _), ..._] => [
+        ADTMetadata(
+          id,
+          List.map(
+            meta => {
+              switch (meta) {
+              | ExceptionMetadata(_, variant, name) => (variant, name)
+              | _ => failwith("impossible by partition")
+              }
+            },
+            exception_meta,
+          ),
+        ),
+        ...non_exception_meta,
+      ]
+    | _ => non_exception_meta
+    };
+
+  let round_to_8 = n => Int.logand(n + 7, Int.lognot(7));
+
+  let buf = Buffer.create(256);
+  Buffer.add_int32_le(buf, 0l); // Slot where pointer to next will live
+  Buffer.add_int32_le(buf, 0l); // Slot where module id will live
+  Buffer.add_int32_le(buf, Int32.of_int(List.length(type_metadata)));
+
+  List.iter(
+    meta => {
+      switch (meta) {
+      | ADTMetadata(id, cstrs) =>
+        let section_length =
+          List.fold_left(
+            (total, (_, cstr)) =>
+              total + 12 + round_to_8(String.length(cstr)),
+            8,
```

Comment:
They're the sizes of headers for the section. Constants? Maybe. It's dependent on the following code, so it didn't really make sense to make them constants. It might be better to just add a comment that links to the doc I wrote about it.

Reply:
Ah, I already put a comment.